[megatron] support mixtral training with megatron backend

ccclyu opened the pull request:
# What does this PR do?

Add support of Mixtral MOE model training with Megatron backend including ``Mixtral8x7B`` and ``Mixtral8X22B``. 

# ChangeLog:

It is still labor-heavy to add new type of model to ``mcore`` format including the following changes:
- ``hf_to_mcore_config_mixtral``: convert `hf_config` to `TransformerConfig`. some common configs are merged into one function `_get_base_transformer_config`. 
- ``MixtralModel`` in model_initialzier.py: implement a model initializer class to initialize GPTModel from config.
- `McoreToHFWeightConverterMixtral`:  model conversion class from mcore to huggingface basically rename
- model entry in `registry.py`: add entry function or class in corresponding registries.  

# Usage

- convert [Mixtral-8x7B-Instruct-v0.1](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1/) to mcore format ([converted](https://huggingface.co/clyu/Mixtral-8x7B-Instruct-v0.1-mcore/tree/main))
- Run RLOO script as follows:

```bash
set -x

train_files=$gsm8k_train_path
test_files=$gsm8k_test_path

export MEGATRON_MODEL="Mixtral-8x7B-Instruct-v0.1-mcore"

python3 -m verl.trainer.main_ppo --config-path=./config --config-name='ppo_megatron_trainer' \
    algorithm.adv_estimator=rloo \
    data.train_files=$train_files \
    data.val_files=$test_files \
    data.train_batch_size=128 \
    data.truncation="left" \
    data.max_prompt_length=512 \
    data.max_response_length=4096 \
    actor_rollout_ref.model.path=Mixtral-8x7B-Instruct-v0.1 \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=128 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.actor.megatron.tensor_model_parallel_size=8 \
    actor_rollout_ref.actor.megatron.pipeline_model_parallel_size=2 \
    actor_rollout_ref.actor.megatron.use_dist_checkpointing=True \
    actor_rollout_ref.actor.megatron.dist_checkpointing_path=${MEGATRON_MODEL} \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=8 \
    actor_rollout_ref.rollout.max_num_batched_tokens=8192 \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.4 \
    actor_rollout_ref.rollout.n=4 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=128 \
    actor_rollout_ref.ref.megatron.tensor_model_parallel_size=8 \
    actor_rollout_ref.ref.megatron.pipeline_model_parallel_size=2 \
    actor_rollout_ref.ref.megatron.use_dist_checkpointing=True \
    actor_rollout_ref.ref.megatron.dist_checkpointing_path=${MEGATRON_MODEL} \
    algorithm.use_kl_in_reward=True \
    algorithm.kl_ctrl.kl_coef=0.001 \
    trainer.critic_warmup=0 \
    trainer.val_before_train=True \
    trainer.logger=['console','wandb'] \
    trainer.log_val_generations=100 \
    trainer.project_name='verl_gsm8k_test' \
    trainer.experiment_name='mixtral-8x7b-rloo-gsm8k' \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=2 \
    trainer.save_freq=50 \
    trainer.test_freq=10 \
    trainer.total_epochs=15
```
- RLOO training can improve GSM8K validation accuracy from 74% to 85%

# What is Missing

- refactor hf2mcore conversion scripts as https://github.com/volcengine/verl/pull/1267
- Have a good design of onboarding new model class to avoid labor-intension changes.  


## Before submitting

- [x] Did you read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide) and finish the [code format check](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting)? 
- [x] Did you make sure to update the documentations with your changes in the [docs](https://github.com/volcengine/verl/tree/main/docs) especially for breaking config etc?
- [ ] Did you write any test cases if necessary? Please add CI tests to your new feature.  

# Additional Info: 
- **Issue Number**: None
- **Training**:  Megatron
- **Inference**:  None

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/volcengine/verl?pullRequest=1325) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/volcengine/verl?pullRequest=1325) before we can accept your contribution.<br/><hr/>**changlyu** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/volcengine/verl?pullRequest=1325) it.</sub>

BearBiscuit05 commented:
Great PR! Please complete the CLA first. Thanks!

eric-haibin-lin reviewed (commented):
please add a record in https://github.com/volcengine/verl/blob/main/docs/experiment/ppo.rst

ccclyu commented:
@ann-qin-lu could you also please review the code? thanks!

ccclyu commented on verl/models/mcore/registry.py:
```diff
@@ -22,35 +28,70 @@
+)
+
+# Registry for model configuration converters
+MODEL_CONFIG_CONVERTER_REGISTRY: Dict[str, Callable[[PretrainedConfig, torch.dtype], TransformerConfig]] = {
```
good suggestion. yes, will use one `enums` for supportedmodel.

ccclyu commented on verl/models/mcore/registry.py:
```diff
@@ -59,51 +100,56 @@ def hf_to_mcore_config(hf_config: PretrainedConfig, dtype: torch.dtype) -> Trans
-        "Qwen2ForCausalLM": McoreToHFWeightConverterDense,
-        "Qwen2MoeForCausalLM": McoreToHFWeightConverterQwen2Moe,
-    }
+def get_mcore_weight_converter(hf_config: PretrainedConfig, dtype: torch.dtype) -> Any:
```
thanks. fixed.

ann-qin-lu commented on verl/models/mcore/config_converter.py:
```diff
@@ -15,86 +16,93 @@
+    # Update with any provided overrides
+    base_config.update(kwargs)
+
+    return TransformerConfig(**base_config)
```
[nit] good to add a debug log here to show the overridden TFConfig.

ann-qin-lu commented on verl/models/mcore/model_forward.py:
```diff
@@ -18,7 +19,8 @@
 
 
-def gptmodel_forward_dense(model, input_ids, attention_mask, position_ids, sequence_parallel, value_model=False, pack_seqs=True):
+def gptmodel_forward(model, input_ids, attention_mask, position_ids, sequence_parallel, value_model=False, pack_seqs=True):
```
nit: add inline comment to show what models have been tested with this method

ann-qin-lu commented on verl/models/mcore/saver.py:
```diff
@@ -453,3 +453,7 @@ def _broadcast_tp_shard_tensor_qkv(tensor, q_name, k_name, v_name, src_pp_rank):
+
+
+def merge_megatron_ckpt_gptmodel_mixtral(wrapped_models, config, dtype, is_value_model=False, tie_word_embeddings=False):
+    raise NotImplementedError("merge_megatron_ckpt_gptmodel_mixtral is not implemented")
```
curious what's the blocker for implementing this function?

ccclyu commented on verl/models/mcore/config_converter.py:
```diff
@@ -15,86 +16,93 @@
+    return _get_base_transformer_config(
+        hf_config=hf_config,
+        dtype=dtype,
         use_cpu_initialization=True,
```
will change to `False` as default.

ccclyu commented on verl/models/mcore/model_initializer.py:
```diff
@@ -14,132 +15,117 @@
+        pass
+
+    def get_rope_scaling_args(self):
+        rope_scaling_args = {}
```
there are more `rope_scaling` args that can be extended in the future. like https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_rope_utils.py

ccclyu commented on verl/models/mcore/config_converter.py:
```diff
@@ -15,86 +16,93 @@
-        gated_linear_unit=True,
+
+    # Base configuration with common parameters
+    base_config = {
```
sure. will re-design config logic in the future PR.

ccclyu commented:
addressed comments from @ann-qin-lu @dataproblems and added mixtral 8x22b instruct results ([wandb](https://api.wandb.ai/links/ppo_dev/sbuiuf2d)) in doc @eric-haibin-lin . Can we now approve&merge? @ETOgaosion  Will refactor the code as proposed in https://github.com/volcengine/verl/issues/1371

ccclyu commented on verl/models/mcore/config_converter.py:
```diff
@@ -15,86 +16,93 @@
+        # Common settings
+        "variable_seq_lengths": True,
+        "masked_softmax_fusion": True,
+        "moe_token_dispatcher_type": "alltoall",
```
https://github.com/NVIDIA/Megatron-LM/blob/10b5c5898e950a4ecc62974ed834163713868921/megatron/core/transformer/transformer_config.py#L1058C1-L1063C18
```

        if self.moe_token_dispatcher_type in ['allgather', 'alltoall_seq']:
            if self.variable_seq_lengths is True:
                raise ValueError(
                    f"Token dispatcher type: {self.moe_token_dispatcher_type} does not support "
                    f"variable sequence length, please use alltoall dispatcher instead."
                )
```

the default value for `moe_token_dispatcher_type` is `allgather`, which causes the above conflict. we set `alltoall` explicitly here.
